error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/CanonicalConstantNaming.java

```diff
@@ -0,0 +1,157 @@
+package tech.picnic.errorprone.bugpatterns;
+
+import static com.google.errorprone.BugPattern.LinkType.CUSTOM;
+import static com.google.errorprone.BugPattern.SeverityLevel.WARNING;
+import static com.google.errorprone.BugPattern.StandardTags.LIKELY_ERROR;
+import static com.google.errorprone.matchers.Matchers.allOf;
+import static com.google.errorprone.matchers.Matchers.hasModifier;
+import static tech.picnic.errorprone.utils.Documentation.BUG_PATTERNS_BASE_URL;
+
+import com.google.auto.service.AutoService;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.ErrorProneFlags;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.bugpatterns.BugChecker.VariableTreeMatcher;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.fixes.SuggestedFixes;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.CompilationUnitTree;
+import com.sun.source.tree.VariableTree;
+import com.sun.source.util.TreeScanner;
+import java.util.Locale;
+import java.util.regex.Pattern;
+import javax.inject.Inject;
+import javax.lang.model.element.Modifier;
+import org.jspecify.annotations.Nullable;
+import tech.picnic.errorprone.utils.Flags;
+
+/**
+ * A {@link BugChecker} that flags constant variables that do not follow the upper snake case naming
+ * convention.
+ *
+ * <p>This check will rewrite the following variables with all its references:
+ *
+ * <pre>{@code
+ * private static final int simpleNumber = 1;
+ * }</pre>
+ *
+ * <p>To the following:
+ *
+ * <pre>{@code
+ * private static final int SIMPLE_NUMBER = 1;
+ * }</pre>
+ *
+ * @apiNote This check has an optional flag `AllowedConstantNames` that represents a list of field
+ *     names to exclude from this check.
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+    summary = "Constant variables should adhere to the `UPPER_SNAKE_CASE` naming convention",
+    link = BUG_PATTERNS_BASE_URL + "CanonicalConstantNaming",
```

Comment:
I think `ConstantNaming` or `ConstantName` would be a better name for this check; the word "canonical" has a different meaning in the context of the other two checks named `Canonical*`. (We could say that it's about _idiomatic_ naming, but that applies to quite some other checks as well.)

Between `ConstantNaming` and `ConstantName` I can make an argument either way; we're not consistent in this respect. Will use the former for now.